### ISSUE TITLE
Plotting wind barbs

### ISSUE THREAD

**Issue by sjoro** [vispy/vispy#1563]
a feature request:
i'm using vispy to plot meteorological products. it would be great to have have a marker or a way to plot wind barbs. this type of visualization is available in [matplotlib](https://matplotlib.org/examples/pylab_examples/barb_demo.html)

**Comment by djhoese** [comment]
Hi @sjoro. I think we had a user a year or so ago bring up wind barbs on [gitter](https://gitter.im/vispy/vispy) or in an issue. I can't find the discussion on github anywhere, but I think someone has done something like this. I feel like @kmuehlbauer was in on the conversation too. I feel like I remember seeing some shader code for it too.

The "best" idea would be to have a custom Markers visual where direction and magnitude could be used inside the shader to draw the barbs. I'm not sure how possible this is. But it would be better than having to generate the individual line segments for each point.

**Comment by kmuehlbauer** [comment]
@sjoro I think this was discussed before, unfortunately I wasn't able to dig this up.

I have the same impression as @djhoese that a custom MarkerVisual is the best way to go. And this would indeed be a nice addition. How many different windbarbs are out there?

**Comment by sjoro** [comment]
thanks for the info!  shame that the old discussion could not be found.
regarding the barbs, the representation depends on the wind speed. so, as individual markers there's many different ones. [https://blog.mapbox.com/mapping-wind-barbs-to-show-speed-and-direction-4d3078add03d](Here) is a good example. would each of these barbs need it's own Marker or could it be build on-fly?

**Comment by kmuehlbauer** [comment]
@sjoro Currently `Marker` just get an array of positions where to plot. You could create different markers for the different wind speed ranges. But this won't work here as you also have wind direction as parameter to take into account.

Nevertheless, it should be possible to also provide wind speed and direction to calculate the barb from those values within the shader. This would need some tweaking in `MarkerVisual`-class but should be possible. We also might consider subclassing MarkerVisual for that purpose. WDYT @djhoese @larsoner? Up for an adventure, @sjoro? PR welcome.

**Comment by sjoro** [comment]
i'm happy to assist in anyway i can, but please keep in mind that i have no openGL programming skills :)

**Comment by kmuehlbauer** [comment]
Means, for a range up to 200 knots we need only 45 different windbarb polygons. Depending on the direction of wind the windbarb polygon just need to be rotated accordingly. For use on the southern hemisphere, we just need to flip on x-axis. Could all be assembled in one dedicated function.

**Comment by sjoro** [comment]
looks very promising! thanks @kmuehlbauer !

**Comment by kmuehlbauer** [comment]
This is probably the quickest way to implement this, without touching shaders. In the example the windbarb is build from segments and rotated. Should be fairly easy to implement, @sjoro.

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import numpy as np
from vispy import app, gloo
from vispy.util.transforms import rotate
from vispy.visuals.collections import PathCollection, PolygonCollection


canvas = app.Canvas(size=(800, 800), show=True, keys='interactive')
gloo.set_viewport(0, 0, canvas.size[0], canvas.size[1])
gloo.set_state("translucent", depth_test=False)

start = np.array([[0., 0., 0.],
                  [-0.05, 0., 0.],
                  [-0.05, 5., 0.],
                  [0.05, 5., 0.]])

flag = np.array([[1.45, -0.5, 0],
                 [0, -1.0, 0]])

fullbarb = np.array([[0, -0.5, 0],
                     [1.45, 0, 0],
                     [1.45, -0.1, 0],
                     [0, -0.6, 0]])

halfbarb = np.array([[0, -0.5, 0],
                     [0.75, -0.25, 0],
                     [0.75, -0.35, 0],
                     [0, -0.6, 0]])

stop = np.array([[0.05, 0., 0.],
                 [0., 0., 0.]])

# build barb from segments
barb = np.concatenate((start, (start[-1]+flag)), axis=0)
barb = np.concatenate((barb, (barb[-1]+fullbarb)), axis=0)
barb = np.concatenate((barb, (barb[-1]+fullbarb)), axis=0)
barb = np.concatenate((barb, (barb[-1]+halfbarb)), axis=0)
barb = np.concatenate((barb, stop), axis=0)

paths = PathCollection("agg", color='shared')
polys = PolygonCollection("raw", color='shared')

for i in np.arange(0, 360, 30):
    # rotate barb
    barbplot = barb.dot(rotate(i, (0., 0., 1.))[0:3,0:3])
    # fix issues with small values
    barbplot[abs(barbplot) < 1e-5] = 0
    polys.append(barbplot / 10., color=(0, 0, 0, 1.0))
    paths.append(barbplot / 10., closed=True, color=(0, 0, 0, 1.0))

paths["linewidth"] = 1.0
paths['viewport'] = 0, 0, 800, 800


@canvas.connect
def on_draw(e):
    gloo.clear('white')
    polys.draw()
    paths.draw()


@canvas.connect
def on_resize(event):
    width, height = event.size
    gloo.set_viewport(0, 0, width, height)
    paths['viewport'] = 0, 0, width, height

app.run()

```
![screenshot_20190124_093111](https://user-images.githubusercontent.com/5821660/51665081-3892c900-1fbb-11e9-8e8b-8ed968a38c4b.png)

**Comment by sjoro** [comment]
looks really good, thanks @kmuehlbauer! do you see any performance issues with this if plotting a vast amount of barbs. our product has ~50k winds.

**Comment by kmuehlbauer** [comment]
@sjoro Can you elaborate a bit further about the application? How do you display ~50k windbarbs.

The first image is 1000 barbs, the second 100000. I can't see any reason to display that much windbarbs.

![screenshot_20190124_111033](https://user-images.githubusercontent.com/5821660/51671341-fff9ec00-1fc8-11e9-819e-86103ff518a7.png)
![screenshot_20190124_111110](https://user-images.githubusercontent.com/5821660/51671346-038d7300-1fc9-11e9-9bce-48f5a580c420.png)

**Comment by sjoro** [comment]
thank you!
95% of the time we inspect the winds per channel, so good performance for 10k winds would be appreciated. officially speaking the flags should be filled, though in my IDL example they are not. so, as a first implementation they would not need to be filled if that brings performance gains. it's more important to be able to just plot them. i'm using vispy to plot other types of products (array data) and need a way to overlay them with winds.

**Comment by rougier** [comment]
Note that you can probably have the rotation made inside the GPU

**Comment by kmuehlbauer** [comment]
@rougier Thanks for chiming in. Actually I'm just reading your great book (http://www.labri.fr/perso/nrougier/python-opengl/#id29). From my experience the performance loss is not the creation of the vertices (rotation) but the vast amount of items put into the collection, But I might miss something.

If we were going the marker-way how would you tackle the creation of these barbs within the shader. I mean we need a line, a triangle, and two arrow-like lines in different combinations. I'm willing to give this a try (would also be a nice addition to glumpy) if you can give me a hand on this.

**Comment by rougier** [comment]
@kmuehlbauer If you barbers are all the same (memory) size, you can add all of them at once with the collection.

I'm not familiar enough with barbs but they are supposed to all look exactly the same, you can consider having a distance field. I explained this in the book with the firefox logo. The other option is to write a specific shader and looking at the barb, it should not be too difficult (lines / triangles). I would start on shadertoy to experiment and if you manage to do it, then you're done.

@sjoro could you explain what is the signification of the triangle, lines, orientations of the barb ? Are they supposed to be parameterized by some variables ?

**Comment by rougier** [comment]
But there is a fixed size of such barbs, right ? In that case a big SDF texture with all the markers might do the trick.

**Comment by kmuehlbauer** [comment]
Yes, the size of the barbs is fixed. You mean, we put all possible variations of these barbs into a SDF texture and select from that texture. I can grasp that this might work, but have no idea how this is coded.

**Comment by rougier** [comment]
There's is a SDF glyph collection that you can use as a starter (or find a font file with barb sign inside).
The new collection would be pretty similar but much simpler because you don't need to align characters.

For the alternative (defining bar through function), the problem is to find a good way to parameterize the shape if don't want to have too much tests (if speed = 0 then draw(barb0) else if speed=10... else if speed=100... ). This will slow down rendering a lot if you do this for some thousand barbs.

**Comment by sjoro** [comment]
regarding the AMV product, it consists of latitudes and longitudes for the location, and then wind speed and direction for each wind barb. it also has the height (i.e. pressure) for each wind that i've used for coloring the barbs. and just to confirm, the size of a barb is fixed, but should be preferably configurable.

**Comment by rougier** [comment]
Size, position, color & rotation are easy. The only difficulty is to draw the shape depending on the speed. Using analytical functions might not be a good option unless we find a way to reduce the number of `if/else` statements.

**Comment by kmuehlbauer** [comment]
@rougier We can break it down to one if/else (for the calm-barb which is a double circle) and three while-loops (flags, fullbarbs, halfbarb). 

If we think about 145 knots there will be calculations for the shaft (1), for two flags (2), for four fullbarbs (4) and one halfbarb (1). This means calculation and mixing of 8 elements for that one barb, where for the different elements just the position is shifted a bit.

No idea, if that's already a showstopper.

**Comment by kmuehlbauer** [comment]
@sjoro @rougier I adapted the gloo-quiver plot using a windbarb calculation which I found on [shadertoy](https://www.shadertoy.com/view/XsKyR1). The wind originates at mouse position and is getting stronger with distance.
![screenshot_20190128_100033](https://user-images.githubusercontent.com/5821660/51824919-e36cf500-22e3-11e9-8876-523abeb6d4c9.png)

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# -----------------------------------------------------------------------------

from vispy import app, gloo

vertex = """
attribute vec2 position;
void main()
{
    gl_Position = vec4(position, 0.0, 1.0);
}
"""

fragment = """
// adapted from https://www.shadertoy.com/view/XsKyR1
#include "math/constants.glsl"
#include "arrows/arrows.glsl"
#include "antialias/antialias.glsl"

uniform vec2 iResolution;
uniform vec2 iMouse;

#define PI	3.1415926535

//---------------------------------------
// Utils


float angle(vec2 p)
{
    return atan(p.y, p.x);
}

float angle(vec2 p0, vec2 p1)
{
    float a =  angle(p1) - angle(p0);
    a = mod(a-PI, 2.*PI)-PI;

    return a;
}


//---------------------------------------
// Segment

struct Segment
{
    vec2 A;
    vec2 B;
};

vec2 getPt(Segment c, float t)
{
    return mix(c.A, c.B, t);
}

float dist(Segment s, vec2 p)
{
    vec2 pa = p-s.A, ba = s.B-s.A;
    float h = clamp( dot(pa,ba)/dot(ba,ba), 0.0, 1.0 );
    return length( pa - ba*h );
}

//---------------------------------------
// Circle

struct Circle
{
    vec2 ctr;
    float rad;
};

//vec2 getPt(Circle c, float t)
//{
//    return c.ctr + c.rad*vec2(cos(t*PI*2.), sin(t*PI*2.));
//}

float dist(Circle c, vec2 p)
{
    return (length(c.ctr-p)-c.rad);
}


//---------------------------------------
// Triangle

struct Tri
{
    vec2 A;
    vec2 B;
    vec2 C;
};

float winding(Tri q,vec2 p)
{
    float w = 0.0;

    w += angle(q.A-p, q.B-p);
    w += angle(q.B-p, q.C-p);
    w += angle(q.C-p, q.A-p);

    return w;
}

bool isInside(Tri q,vec2 p)
{
    float w = winding(q, p);
    return abs(w)>0.001;
}

float dist(Tri q, vec2 p)
{
    float d = min(min(dist(Segment(q.A, q.B), p), 
                      dist(Segment(q.B, q.C), p)),  
                  dist(Segment(q.C, q.A), p));
    if(isInside(q,p))
        d *= -1.;
    return d;
}

//---------------------------------------
// Drawing Macros

// draw the inside of the shape
#define DRAW(O,P,C,CC)	CC=mix(CC,C,mix(1.,0.,clamp(dist(O,P)*iResolution.x*2., -1., 1.)*.5+.5))

// draw the contour of the shape 
#define DRAW_C(O,P,C,W,CC)	CC=mix(CC,C,mix(1.,0.,clamp(abs(dist(O,P))*iResolution.x*2.-W, -1., 1.)*.5+.5))

// draw the contour of the shape with alpha
#define DRAW_CA(O,P,C,W, CC)	CC=mix(CC,C.rgb,(mix(1.,0.,clamp(abs(dist(O,P))*iResolution.y/2. -  W, -1., 1.)*.5+.5))*(1.-C.a))


#define NB	16.

void main()
{
    const float M_PI = 3.14159265358979323846;
    const float SQRT_2 = 1.4142135623730951;
    const float linewidth = 3.0;
    const float antialias =  1.0;

    const float rows = NB;
    const float cols = NB;
    
    vec2 texcoord = gl_FragCoord.xy;
    vec2 size   = iResolution.xy / vec2(cols,rows);
    vec2 center = (floor(texcoord / size) + vec2(0.5, 0.5)) * size;
    
    vec2 uv = texcoord / iResolution.xy;
    vec2 uv_ = floor(uv * NB) / NB;
    vec2 s = (uv_ - (iMouse - center)); 

    float dx = 0.5 / NB;
    vec2 O = vec2(dx);
    vec2 X = normalize(s) * dx / SQRT_2;
    vec2 Y = X.yx * vec2(-1., 1.);
    vec2 uv_c = (uv-uv_);

    // caclulate barb items
    float speed = length(s) / 8;
    int flag = int(floor(speed / 50.));
    speed -= float (50 * flag);
    int longbarb = int(floor(speed / 10.));
    speed -= float (longbarb * 10);
    int shortbarb = int(floor(speed / 5.));
    int calm = shortbarb + longbarb + flag;
    
    // start color
    vec3 col = vec3(1.);
    
    // calm, plot circles
    if (calm == 0)
    {
        DRAW_C(Circle(O, dx*.2),uv_c,vec3(0.1),1.,col);
        DRAW_C(Circle(O, dx*.1),uv_c,vec3(0.1),.8,col);
    }
    else
    {
        // plot shaft
        DRAW_C(Segment(O, O-X),uv_c,vec3(0.1),1.,col);
        float pos = 1.0;

        // plot flag(s)
        while(flag >= 1)
        {
            DRAW(Tri(O-X*pos, O-X*pos-X*.4-Y*.4, O-X*pos-X*.4),uv_c,vec3(0.1),col);
            flag -= 1;
            pos -= 0.15;
        }
        // plot longbarb(s)
        while(longbarb >= 1)
        {
            DRAW_C(Segment(O-X*pos, O-X*pos-X*.3-Y*.3),uv_c,vec3(0.1),1.,col);
            longbarb -= 1;
            pos -= 0.15;
        }
        // plot shortbarb
        while(shortbarb >= 1)
        {
            if (pos == 1.0)
                pos -= 0.15;
            DRAW_C(Segment(O-X*pos, O-X*pos-X*.15-Y*.15),uv_c,vec3(0.1),1.,col);
            shortbarb -= 1;
            pos -= 0.15;
        }
    }
    gl_FragColor = vec4(col, 1.);
}
"""

canvas = app.Canvas(size=(2*512, 2*512), keys='interactive')
canvas.context.set_state(blend=True, 
                         blend_func=('src_alpha', 'one_minus_src_alpha'),
                         blend_equation='func_add')


@canvas.connect
def on_draw(event):
    gloo.clear('white')
    program.draw('triangle_strip')


@canvas.connect
def on_resize(event):
    program["iResolution"] = event.size
    gloo.set_viewport(0, 0, event.size[0], event.size[1])


@canvas.connect
def on_mouse_move(event):
    x, y = event.pos
    program["iMouse"] = x, canvas.size[1] - y
    canvas.update()


program = gloo.Program(vertex, fragment, count=4)
dx, dy = 1, 1
program['position'] = (-dx, -dy), (-dx, +dy), (+dx, -dy), (+dx, +dy)
program["iMouse"] = (0., 0.)

if __name__ == '__main__':
    canvas.show()
    app.run()

```

**Comment by kmuehlbauer** [comment]
Below is the main-function after dissolving the DRAW-definitions. Variable `d` should now resemble something like a distance-field but with linewidth and kind of aliasing already applied. It might be possible to make this more efficient.

@rougier Is this the analytical solution you mentioned? How do we get this into VisPy now? Via `MarkerVisual` or via kind of `ArrowVisual`? What do you suggest? We would also need to feed the windspeed and direction into it.

```python
void main()
{
    const float M_PI = 3.14159265358979323846;
    const float SQRT_2 = 1.4142135623730951;
    const float linewidth = 1.0;
    const float aa_factor =  2.0;

    const float rows = NB;
    const float cols = NB;
    
    vec2 texcoord = gl_FragCoord.xy;
    vec2 size   = iResolution.xy / vec2(cols,rows);
    vec2 center = (floor(texcoord / size) + vec2(0.5, 0.5)) * size;
    
    vec2 uv = texcoord / iResolution.xy;
    vec2 uv_ = floor(uv * NB) / NB;
    vec2 s = (uv_ - (iMouse - center)); 

    float dx = 0.5 / NB;
    vec2 O = vec2(dx);
    vec2 X = normalize(s) * dx / SQRT_2;
    vec2 Y = X.yx * vec2(-1., 1.);
    vec2 uv_c = (uv-uv_);

    // calculate barb items
    float speed = length(s) / 8;
    int flag = int(floor(speed / 50.));
    speed -= float (50 * flag);
    int longbarb = int(floor(speed / 10.));
    speed -= float (longbarb * 10);
    int shortbarb = int(floor(speed / 5.));
    int calm = shortbarb + longbarb + flag;
    
    // starting distance
    float d = 1.0;
    
    // calm, plot circles
    if (calm == 0)
    {
        d = min(d, clamp(abs(dist(Circle(O, dx * .1),uv_c))*iResolution.x*aa_factor - linewidth, -1., 1.) * .5 + .5);
        d = min(d, clamp(abs(dist(Circle(O, dx * .2),uv_c))*iResolution.x*aa_factor - linewidth, -1., 1.) * .5 + .5);
    }
    else
    {
        // plot shaft
        d = min(d, clamp(abs(dist(Segment(O, O-X),uv_c))*iResolution.x*aa_factor - linewidth, -1., 1.) * .5 + .5);
        float pos = 1.0;

        // plot flag(s)
        while(flag >= 1)
        {
            d = min(d, clamp(dist(Tri(O-X*pos, O-X*pos-X*.4-Y*.4, O-X*pos-X*.4),uv_c)*iResolution.x*aa_factor, -1., 1.)*0.5 + 0.5);
            d = min(d, clamp(dist(Tri(O-X*pos, O-X*pos-X*.4-Y*.4, O-X*pos-X*.4),uv_c)*iResolution.x*aa_factor - linewidth, -1., 1.)*0.5 + 0.5);
            flag -= 1;
            pos -= 0.15;
        }
        // plot longbarb(s)
        while(longbarb >= 1)
        {
            d = min(d, clamp(abs(dist(Segment(O-X*pos, O-X*pos-X*.3-Y*.3),uv_c))*iResolution.x*aa_factor - linewidth, -1., 1.) * .5 + .5);
            longbarb -= 1;
            pos -= 0.15;
        }
        // plot shortbarb
        while(shortbarb >= 1)
        {
            if (pos == 1.0)
                pos -= 0.15;
            d = min(d, clamp(abs(dist(Segment(O-X*pos, O-X*pos-X*.15-Y*.15),uv_c))*iResolution.x*aa_factor - linewidth, -1., 1.) * .5 + .5);
            shortbarb -= 1;
            pos -= 0.15;
        }
    }
    gl_FragColor = vec4(vec3(d), 1);
}
```

**Comment by rougier** [comment]
Probably via Marker visual (or a subclass). Orientation and speed needs to be fed via attributes.

**Comment by kmuehlbauer** [comment]
I thought so. But I have a hard time to understand the difference between plotting the barb at some coordinate and the momentary code, where everything depends on the `iResolution` and the artificial grid within the shader. Can you shed some light or point me to some examples?

**Comment by kmuehlbauer** [comment]
@rougier I think I've resolved my problems in understanding. 

@sjoro I'm currently thinking a bit about which parameters to feed to the `WindbarbVisual`. We might think about feeding u/v wind components. This would be easy in the shader to get direction and length of that vector. We can feed speed and direction also. Or we can make this both available. Any more ideas?

**Comment by kmuehlbauer** [comment]
OK, here comes first iteration of `WindbarbVisual`including application. Using 50k barbs I get roughly 2 FPS while updating everything (position, wind, colors) on my machine. Using 10k barbs this gets up to decent 12 FPS. If you zoom in this accelerates even more depending on the zoom ratio.
![screenshot_20190130_143112](https://user-images.githubusercontent.com/5821660/51984487-b82d0600-249b-11e9-96ef-e951c53c2597.png)

The source code is here in the [windbarb-branch[(https://github.com/kmuehlbauer/vispy/blob/windbarb/examples/basics/visuals/windbarbs.py)

There are two implementations of the signed-segment-distance and the sdf-triangle. @rougier your implementation of the sdf-triangle is a bit faster than the one from shadertoy whereas for the signed segment distance this is the other way round.

**Comment by sjoro** [comment]
@kmuehlbauer this looks great! thanks for your efforts. regarding the input parameters for `WindbarbVisual`, we have speed and direction in our products, so that's what i would prefer. also, the speed should be fed (and the barbs drawn) in SI-units, i.e., m/s. if the user wants to use knots or km/h, they should maybe do the conversion themselves. alternatively there could be an optional keyword for units.

**Comment by kmuehlbauer** [comment]
@sjoro Can you explain the SI-units part? You mean you want to feed m/s?

What do you mean with **and the barbs drand**?
> also, the speed should be fed (and the barbs drawn) in SI-units, i.e., m/s.

**Comment by sjoro** [comment]
yes, the input unit should be in m/s. so, given the input of 10 to `WindbarbVisual` it should draw two full barbs. 10 knots (= 5 m/s) would be one full barb. hope this helps.

**Comment by sjoro** [comment]
[Here](https://www.wmo.int/pages/prog/www/ISS/Meetings/IPET-DRMM_Beijing2015/Documents/IPET-DRMM-III_Doc11-5_GDPFS_Manual_Appendix_II-4.pdf)'s the instructions for drawing barbs from World Meteorological Organization (WMO)
> True direction, in tens of degrees, from which wind is blowing and wind speed in units indicated by
Wind is represented by barbs and solid pennants in black, the full barbs representing 5 m s -1 or 10 knots, the half barbs representing 2.5 m s -1 or 5 knots and the solid pennant representing 25 m s -1 or 50 knots.
The wind shaft in black is directed along the axis of the wind towards the centre of the station circle and
stops at its circumference. All pennants and barbs lie to the left of the wind shaft in the northern hemisphere and to the right of the wind shaft in the southern hemisphere. Barbs are at an angle of approximately 120° from the wind shaft. Pennants are triangles with their bases on the wind shaft.
A calm should be indicated by a circle drawn around the station circle:
Missing wind speed should be indicated by placing an “ x” at the end of the wind shaft in lieu of the wind barbs. Wind direction is indicated in the usual manner, e.g. x——o. When the wind direction is missing, no wind should be plotted.

**Comment by kmuehlbauer** [comment]
So you want the following:

```bash
 2.5 m/s   one half barb
 5.0 m/s   one full barb
10.0 m/s   two full barbs
25.0 m/s   one flag
50.0 m/s   two flags
```

**Comment by sjoro** [comment]
that's my preference, yes. people used to knots probably disagree :)

**Comment by kmuehlbauer** [comment]
No problem, it's just a factor of two :grinning:

**Comment by kmuehlbauer** [comment]
@sjoro Would be nice if you could test/review https://github.com/kmuehlbauer/vispy/blob/windbarb/examples/basics/visuals/windbarbs.py.

**Comment by kmuehlbauer** [comment]
implemented in #1567
